### PR TITLE
[REF] discuss: call tests with mocked network layer

### PR DESCRIPTION
**Pull request by ThanhDodeurOdoo**
- Add a Mock for the network layer of calls (`rtcService.network`),
which allows for the testing of features that depend on network events,
without using online tests (which would require local network access in
the test environment).

- Improve the stream testing API, so tests for stream
cleanup are also added, which guards against memory leaks that could
be caused if streams aren't properly closed and the associated lingering
OS/browser level media indicators (eg: webcam activity light).

- Add a test for https://github.com/odoo/odoo/pull/230287 now that it's
possible to test such features.

### CODE REVIEW

File: addons/mail/static/tests/discuss/call/call.test.js [hunk 1]
```diff
@@ -823,3 +826,107 @@ test("dynamic focus switches to talking participant", async () => {
...
     await contains(".o-dropdown-item:contains('Autofocus speaker')");
 });
+
+test("should not leave unclosed streams after leaving a call", async () => {
+    const pyEnv = await startServer();
+    const channelId = pyEnv["discuss.channel"].create({ name: "General" });
+    await start();
+
+    await openDiscuss(channelId);
+    await click("[title='Start Call']");
```
Comment (alexkuhn) [inline addons/mail/static/tests/discuss/call/call.test.js:897]: Nitpick: I don't like seeing arbitrary newlines inside test.

When I scan the file these newlines give my brain wrong impression these are splits between tests when they aren't.

From experience with writing tests and adapting guidelines to make them more maintainable: each test block should have no empty new lines.

Reason is that this makes it easier to see each block of test as cohesive bits, and makes it easier to see which LOC is on which test. In production code we avoid that i.e. by putting classes in different files, but it's a bit extreme to put each test in its own file.
And for methods of classes that is ok to have empty newlines: methods of a class are coupled by being in the same class, which is not the same as tests in a single HOOT file, hence why this is not the same.

Another argument against newline: this incites to make more LOCs in tests than necessary, which is not good because tests require maintenance and we should be cautious with not adding more LOCs than necessary.

File: addons/mail/static/tests/discuss/call/call.test.js [hunk 2]
```diff
@@ -823,3 +826,107 @@ test("dynamic focus switches to talking participant", async () => {
...
+    expect(screenStream.getTracks()[0].readyState).toBe("live");
+    expect(streams.length).toBe(4); // in total we requested 1 audio, 2 camera, 1 screen
+    pyEnv["bus.bus"]._sendone(serverState.partnerId, "mail.record/insert", {
+        "discuss.channel": [
+            {
+                id: channelId,
+                rtc_session_ids: [],
+            },
+        ],
+    });
```
Comment (alexkuhn) [inline addons/mail/static/tests/discuss/call/call.test.js]: Should put comment this simulates someone ending the call.
Or better if we can: invoke the proper RPC to do that with the appropriate user.

File: addons/mail/static/tests/mail_test_helpers.js [hunk 3]
```diff
@@ -489,28 +484,105 @@ export function mockGetMedia() {
...
+ * Mocks {import("@mail/discuss/call/common/rtc_service").Network} and allows testing of features that rely on network behavior, such
+ * as other participants changing the state of their microphone, sharing screen,...
+ *
+ * @param {string} channelName
+ * @returns {Promise<MockNetwork>}
+ */
+export async function mockRtcNetwork({ channelName }) {
+    const pyEnv = await startServer();
+    const channelId = pyEnv["discuss.channel"].create({ name: channelName });
+    const mockNetwork = new EventTarget();
```
Comment (alexkuhn) [inline addons/mail/static/tests/mail_test_helpers.js]: `mockRtcNetwork` should not have concern to create a `discuss.channel`.
This makes it hard to understand each test because this relies on shared data in the helper, and also if the helper is not complete enough for newer test then changing it would impact older tests, which is a big issue.

When we write some tests, ideally we should not touch them again unless there's a failure. A test should not mutate it's setup magically after we wrote it once: each change should be quite explicit.

If we don't do that, then we're gonna have a hard time to understand the relevancy of these old tests, and it's gonna be a big waste of time because the shared helper create far too much data that is actually useful for each of these tests. And also adapting the tests require to work on all tests at once which is painful.

------

What we should do is have test setup the `discuss.channel` explicitly and the mock function can have a channel being passed as parameter to make the appropriate mocks.

File: addons/mail/static/tests/discuss/call/call.test.js (hunk 2 above)
Comment (ThanhDodeurOdoo) [reply addons/mail/static/tests/discuss/call/call.test.js]: added a comment, it represents any unlink not necessarily caused by the client (for example it could be an admin going to the technical view of rtc session and clicking the unlink buttom from the form view), the flow where the user cleanly disconects with the UI interaction is in the other test that does the base case

File: addons/mail/static/tests/mail_test_helpers.js [hunk 4]
```diff
@@ -489,28 +484,105 @@ export function mockGetMedia() {
...
+ * Mocks {import("@mail/discuss/call/common/rtc_service").Network} and allows testing of features that rely on network behavior, such
+ * as other participants changing the state of their microphone, sharing screen,...
+ *
+ * @param {Object} param0
+ * @param {MockServerEnvironment} param0.pyEnv
+ * @param {Partial<OdooEnv>} param0.env
+ * @param {number} param0.channelId
+ * @returns {Promise<MockNetwork>}
+ */
+export async function mockRtcNetwork({ pyEnv, env, channelId }) {
```
Comment (alexkuhn) [inline addons/mail/static/tests/mail_test_helpers.js]: No need to pass `pyEnv`: can deduce from `MockServer.current.env`.

File: addons/mail/static/tests/mail_test_helpers.js [hunk 5]
```diff
@@ -489,28 +484,105 @@ export function mockGetMedia() {
...
+    return {
+        channelId,
+        mockRemote(partnerName) {
+            const sessionId = pyEnv["discuss.channel.rtc.session"].create({
+                channel_member_id: pyEnv["discuss.channel.member"].create({
+                    channel_id: channelId,
+                    partner_id: pyEnv["res.partner"].create({ name: partnerName }),
+                }),
+                channel_id: channelId,
+            });
```
Comment (alexkuhn) [inline addons/mail/static/tests/mail_test_helpers.js]: I would also not have `discuss.channel.member` being created by this helper: the channel member should be explicitly created in the tests, and `mockRemote()` should have `memberId` as param.

File: addons/mail/static/tests/mail_test_helpers.js [hunk 6]
```diff
@@ -489,28 +484,105 @@ export function mockGetMedia() {
...
+            return {
+                sessionId,
+                async updateState(state) {
```
Comment (alexkuhn) [inline addons/mail/static/tests/mail_test_helpers.js]: Would rename to `updateConnectionState()` for clarity compared to `updateInfo` and `updateUpload`, that we know this is low-level change of the connection state of the remote.

File: addons/mail/static/tests/discuss/call/call.test.js [hunk 7]
```diff
@@ -859,3 +862,111 @@ test("should not show context menu on participant card when not in a call", asyn
...
+    expect(screenStream.getTracks()[0].readyState).toBe("live");
+    expect(streams.length).toBe(4); // in total we requested 1 audio, 2 camera, 1 screen
+    // represents the unlink of a rtc session
+    pyEnv["bus.bus"]._sendone(serverState.partnerId, "mail.record/insert", {
+        "discuss.channel": [
+            {
+                id: channelId,
+                rtc_session_ids: [],
+            },
+        ],
```
Comment (alexkuhn) [inline addons/mail/static/tests/discuss/call/call.test.js]: Would be best to `pyEnv["discuss.channel.rtc.session"].unlink()` and add the `bus._sendmany()` in the side-effect of unlink of `discuss_channel_rtc_session.js`.

A bit like the `create` function of same file: https://github.com/odoo/odoo/blob/19.0/addons/mail/static/tests/mock_server/mock_models/discuss_channel_rtc_session.js#L54

File: addons/mail/static/tests/discuss/call/call.test.js [hunk 8]
```diff
@@ -859,3 +862,111 @@ test("should not show context menu on participant card when not in a call", asyn
     );
     await contains(".o-discuss-CallContextMenu");
 });
+
+test("should not leave unclosed streams after leaving a call", async () => {
```
Comment (alexkuhn) [inline addons/mail/static/tests/discuss/call/call.test.js]: Better word the test in a positive like how the assertions are made.
Positive statements are easier to test and are more relevant functionally.
Something like `"Leaving a call should close all the streams"`.

File: addons/mail/static/tests/discuss/call/call.test.js [hunk 9]
```diff
@@ -859,3 +862,111 @@ test("should not show context menu on participant card when not in a call", asyn
...
+    expect(streams[2].getTracks()[0].readyState).toBe("live");
+    await triggerEvents(".o-discuss-Call-mainCards", ["mousemove"]); // show overlay
+    await click(".o-discuss-CallActionList button[aria-label='Disconnect']");
+    await contains(".o-discuss-Call", { count: 0 });
+    expect(streams[0].getTracks()[0].readyState).toBe("ended");
+    expect(streams[1].getTracks()[0].readyState).toBe("ended");
+    expect(streams[2].getTracks()[0].readyState).toBe("ended");
+});
+
+test("all streams are properly closed when requesting new ones and when abruptly disconnected", async () => {
```
Comment (alexkuhn) [inline addons/mail/static/tests/discuss/call/call.test.js]: Could be split in 2 tests:
- `Requesting new streams closes old ones`
- `Abrupt disconnection should close all streams`
That would make each test simpler and easier to get what they do.

File: addons/mail/static/tests/discuss/call/call.test.js [hunk 10]
```diff
@@ -859,3 +862,118 @@ test("should not show context menu on participant card when not in a call", asyn
...
+    await bobRemote.updateConnectionState("connecting");
+    await contains(".o-discuss-CallParticipantCard[title='Bob'] .fa-exclamation-triangle");
+    await bobRemote.updateConnectionState("connected");
+    await contains("span[data-connection-state='connected']");
+    await bobRemote.updateUpload("screen", createVideoStream().getVideoTracks()[0]);
+    await contains(".o-discuss-CallParticipantCard[title='Bob'] video");
+    await bobRemote.updateInfo({ isRaisingHand: true });
+    await contains(".o-discuss-CallParticipantCard[title='Bob'] .fa-hand-paper-o");
+    await contains(".o-discuss-Call-notification:contains('Bob raised their hand')");
+});
```
Comment (alexkuhn) [inline addons/mail/static/tests/discuss/call/call.test.js]: Test name is too generic, prefer small test that are business relevant for discuss calls.
Like:
- `Shows connecting state on cards`
- `Can see raised hands from other call participants`

File: addons/mail/static/tests/discuss/call/call.test.js [hunk 11]
```diff
@@ -859,3 +862,118 @@ test("should not show context menu on participant card when not in a call", asyn
...
+    await contains(".o-discuss-CallParticipantCard-avatar", { count: 2 });
+    await streamerRemote.updateUpload("screen", createVideoStream().getVideoTracks()[0]);
+    await contains(".o-discuss-CallParticipantCard-avatar", { count: 2 });
+    await contains(".o-discuss-CallParticipantCard video", { count: 1 });
+    await click(".o-discuss-CallParticipantCard[title='Streamer'] video");
+    await contains(".o-discuss-CallParticipantCard-avatar", { count: 1 });
+    await contains(".o-discuss-CallParticipantCard video", { count: 1 });
+    await streamerRemote.updateUpload("screen", null);
+    await contains(".o-discuss-CallParticipantCard-avatar", { count: 2 });
+});
```
Comment (alexkuhn) [inline addons/mail/static/tests/discuss/call/call.test.js]: Would name `"show all participants on other user stops screen share"` like the original PR that fixes it. This wording is IMO more understandable than "grid view on share screen defocus", as we don't use words "restore" / "grid view" / "defocus", in addition to name of test being unclear that the screen share change doesn't come from current user but actually the one screen-sharing just stopped the stream from their end.

**Review by alexkuhn — COMMENTED** [review]
By the way all `{ count: 1 }` can be removed: this is implicit and it's easier to read the assertion without them :)

**Review by alexkuhn — COMMENTED** [review]
LGTM otherwise

File: addons/mail/static/tests/discuss/call/call.test.js [hunk 12]
```diff
@@ -859,3 +862,188 @@ test("should not show context menu on participant card when not in a call", asyn
...
+    await click("[title='Join Call']");
+    await contains(".o-discuss-CallParticipantCard[title='Bob']");
+    await bobRemote.updateConnectionState("connected");
+    await bobRemote.updateUpload("screen", createVideoStream().getVideoTracks()[0]);
+    await contains(".o-discuss-CallParticipantCard[title='Bob'] video");
+});
+
+test("network features", async () => {
+    const pyEnv = await startServer();
+    const channelId = pyEnv["discuss.channel"].create({ name: "General" });
```
Comment (alexkuhn) [inline addons/mail/static/tests/discuss/call/call.test.js]: Oops, test redundant with split in 2 tests.

File: addons/mail/static/tests/mail_test_helpers.js [hunk 13]
```diff
@@ -489,28 +484,103 @@ export function mockGetMedia() {
...
+    const rtcServiceIsListening = new Deferred();
+    patchWithCleanup(Network.prototype, {
+        addEventListener(name, f) {
+            if (name === "update") {
+                rtcServiceIsListening.resolve();
+                // disabling the p2p network so that it does not try to send webRTC events like candidates and offers.
+                rtc.network.p2p.disconnect();
+            }
+            mockNetwork.addEventListener(name, f);
+        },
```
Comment (alexkuhn) [inline addons/mail/static/tests/mail_test_helpers.js:539]: Missing `removeEventListener`

File: addons/mail/static/tests/mail_test_helpers.js [hunk 14]
```diff
@@ -489,28 +484,103 @@ export function mockGetMedia() {
...
+                // disabling the p2p network so that it does not try to send webRTC events like candidates and offers.
+                rtc.network.p2p.disconnect();
+            }
+            mockNetwork.addEventListener(name, f);
+            after(() => mockNetwork.removeEventListener(name, f));
+        },
     });
+
+    return {
+        channelId,
```
Comment (alexkuhn) [inline addons/mail/static/tests/mail_test_helpers.js]: `network.channelId` is useless: already provided in param.

**Review by alexkuhn — APPROVED** [review]
@robodoo delegate+ r+
